dev/tests/integration/testsuite/Magento/Downloadable/Model/Observer/SetLinkStatusObserverTest.php

```diff
@@ -0,0 +1,170 @@
+<?php
+
+namespace Magento\Downloadable\Model\Observer;
+
+/**
+ * Integration test for case, when customer is able to download
+ * downloadable product, after order was canceled.
+ */
+class SetLinkStatusObserverTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @var \Magento\Framework\ObjectManagerInterface
+     */
+    protected $objectManager;
+
+    /**
+     * Order collection
+     * @var \Magento\Sales\Model\ResourceModel\Order\Item\CollectionFactory
+     */
+    protected $orderItemFactory;
+
+    /**
+     * @var \Magento\Sales\Model\OrderRepository
+     */
+    protected $orderRepository;
+
+    /**
+     * Initialization of dependencies
+     */
+    protected function setUp()
+    {
+        $this->objectManager = \Magento\TestFramework\Helper\Bootstrap::getObjectManager();
+        $this->orderItemFactory = $this->objectManager->create(
+            \Magento\Sales\Model\ResourceModel\Order\Item\CollectionFactory::class
+        );
+        $this->orderRepository = $this->objectManager->get(
+            \Magento\Sales\Model\OrderRepository::class
+        );
+    }
+
+    /**
+     * Asserting, that links status is expired after canceling of order.
+     *
+     * @magentoAppIsolation enabled
+     * @magentoDbIsolation enabled
+     * @magentoDataFixture Magento/Downloadable/_files/product_downloadable_with_files.php
+     */
+    public function testCheckStatusOnOrderCancel()
+    {
+        /** @var \Magento\Sales\Model\Order $order */
+        $order = $this->prepareOrder();
+        $order = $this->orderRepository->save($order);
+        if ($orderItems = $order->getAllItems()) {
```

Comment:
Looks like assertions will be unreachable in case if order would not have items, therefore making the test green false-positively.